File: src/domain/quincena/balance.ts
/**
 * Restante por quincena y sobrante del mes.
 *
 * Regla (viene directo de tu Excel): el restante de una quincena es
 * ingresos - gastos de esa quincena, SIN filtrar por si ya estan pagados
 * (el checkbox "Ready" es seguimiento aparte, no cambia la matematica).
 * "Cancelado" si se excluye: un gasto cancelado nunca debio contar.
 *
 * "Sobrante del mes" = suma de los dos restantes. Verificado contra tu
 * hoja: 1.215.000 + 1.167.006 = 2.382.006.
 */
import { quincenaKey } from './quincena';
import type { QuincenaKey, Transaction } from '../types';

export interface QuincenaBalance {
  key: QuincenaKey;
  income: number;
  expense: number;
  restante: number;
}

export interface MonthBalance {
  year: number;
  month: number;
  income: number;
  expense: number;
  sobrante: number;
  quincenas: [QuincenaBalance, QuincenaBalance];
}

/**
 * transactions ya debe traer, para cada una, la quincena resuelta
 * (tx.quincenaKey manual, o la calculada con calculateQuincena en la capa
 * que lee de la base de datos). Este archivo no importa calculateQuincena
 * para no acoplar "sumar" con "calcular fecha" — se prueban por separado.
 */
export function calculateQuincenaBalance(
  transactionsWithKey: Array<Transaction & { resolvedQuincenaKey: QuincenaKey }>,
  key: QuincenaKey,
): QuincenaBalance {
  let income = 0;
  let expense = 0;
  for (const tx of transactionsWithKey) {
    if (tx.status === 'cancelled') continue;
    if (tx.resolvedQuincenaKey !== key) continue;
    if (tx.type === 'income') income += tx.amount;
    else expense += tx.amount;
  }
  return { key, income, expense, restante: income - expense };
}

export function calculateMonthBalance(
  transactionsWithKey: Array<Transaction & { resolvedQuincenaKey: QuincenaKey }>,
  year: number,
  month: number,
): MonthBalance {
  const [k1, k2] = [quincenaKey(year, month, 1), quincenaKey(year, month, 2)];
  const q1 = calculateQuincenaBalance(transactionsWithKey, k1);
  const q2 = calculateQuincenaBalance(transactionsWithKey, k2);
  return {
    year,
    month,
    income: q1.income + q2.income,
    expense: q1.expense + q2.expense,
    sobrante: q1.restante + q2.restante,
    quincenas: [q1, q2],
  };
}

File: mobile/lib/domain/quincena/balance.dart
/// Restante por quincena y sobrante del mes.
///
/// Regla (viene directo del Excel original): el restante de una quincena es
/// ingresos - gastos de esa quincena, SIN filtrar por si ya están pagados
/// (el check de "pagado" es seguimiento aparte, no cambia la matemática).
/// "Cancelado" sí se excluye: un gasto cancelado nunca debió contar.
///
/// "Sobrante del mes" = suma de los dos restantes.
library;

import '../types.dart';
import 'quincena.dart';

/// Una transacción con su quincena ya resuelta. Existe para que este
/// archivo se pueda probar con fixtures sin depender de calculateQuincena:
/// "sumar" y "calcular a qué quincena pertenece" se prueban por separado.
class ResolvedTx {
  const ResolvedTx(this.tx, this.key);
  final Transaction tx;
  final QuincenaKey key;
}

List<ResolvedTx> withResolvedQuincena(
  List<Transaction> transactions, [
  List<int> startDays = quincenaPorDefecto,
]) =>
    transactions.map((tx) => ResolvedTx(tx, resolveQuincenaKey(tx, startDays))).toList();

class QuincenaBalance {
  const QuincenaBalance({
    required this.key,
    required this.income,
    required this.expense,
  });

  final QuincenaKey key;
  final int income;
  final int expense;

  int get restante => income - expense;
}

class MonthBalance {
  const MonthBalance({
    required this.year,
    required this.month,
    required this.quincenas,
  });

  final int year;
  final int month;
  final List<QuincenaBalance> quincenas;

  int get income => quincenas.fold(0, (a, q) => a + q.income);
  int get expense => quincenas.fold(0, (a, q) => a + q.expense);
  int get sobrante => quincenas.fold(0, (a, q) => a + q.restante);
}

QuincenaBalance calculateQuincenaBalance(List<ResolvedTx> resueltas, QuincenaKey key) {
  var income = 0;
  var expense = 0;
  for (final r in resueltas) {
    if (r.tx.isCancelled) continue;
    if (r.key != key) continue;
    if (r.tx.isIncome) {
      income += r.tx.amount;
    } else {
      expense += r.tx.amount;
    }
  }
  return QuincenaBalance(key: key, income: income, expense: expense);
}

MonthBalance calculateMonthBalance(List<ResolvedTx> resueltas, int year, int month) => MonthBalance(
      year: year,
      month: month,
      quincenas: quincenasOfMonth(year, month)
          .map((k) => calculateQuincenaBalance(resueltas, k))
          .toList(),
    );
